packages/devextreme/js/__internal/ui/calendar/m_calendar.ts

```diff
@@ -251,116 +261,134 @@ class Calendar<
+    } else if (!this._view.isDateDisabled(currentDate)) {
+      const value = this._updateTimeComponent(currentDate);
       this._selectionStrategy.selectValue(value, e);
     }
   }
 
-  _getSerializationFormat(optionName) {
-    const value = this.option(optionName || 'value');
+  _getSerializationFormat(optionName = 'value'): string | undefined | null {
+    const value = this.option(optionName);
+    const { dateSerializationFormat } = this.option();
 
-    if (this.option('dateSerializationFormat')) {
-      return this.option('dateSerializationFormat');
+    if (dateSerializationFormat) {
+      return dateSerializationFormat;
     }
 
     if (isNumeric(value)) {
       return 'number';
     }
 
     if (!isString(value) || value === '') {
-      return;
+      return undefined;
     }
 
     return dateSerialization.getDateSerializationFormat(value);
   }
 
-  _convertToDate(value) {
+  _convertToDate(value): Date | null {
     return dateSerialization.deserializeDate(value);
   }
 
-  _dateValue(value, event) {
+  _dateValue(value: Date | Date[], event: DxEvent): void {
     if (event) {
       if (event.type === 'keydown') {
         const cellElement = this._view._getContouredCell().get(0);
         event.target = cellElement;
       }
+      // @ts-expect-error ts-error
       this._saveValueChangeEvent(event);
     }
-    this._dateOption('value', value);
+    this._setDateOption('value', value);
   }
 
-  _dateOption(optionName, optionValue) {
-    const isArray = optionName === 'value' && !this._isSingleMode();
-    const value = this.option('value');
-
-    if (arguments.length === 1) {
-      return isArray
-        // @ts-expect-error ts-error
-        ? (value ?? []).map((value) => this._convertToDate(value))
-        : this._convertToDate(this.option(optionName));
-    }
-
+  _setDateOption(
+    optionName: 'value' | 'min' | 'max',
+    optionValue: DateLike | DateLike[],
+  ): void {
+    const isArray = optionName === 'value' && !this._isSingleMode() && Array.isArray(optionValue);
     const serializationFormat = this._getSerializationFormat(optionName);
     const serializedValue = isArray
-      ? optionValue?.map((value) => dateSerialization.serializeDate(value, serializationFormat)) || []
+      ? optionValue?.map((value) => dateSerialization.serializeDate(value, serializationFormat))
+        || []
       : dateSerialization.serializeDate(optionValue, serializationFormat);
 
     this.option(optionName, serializedValue);
   }
 
+  _getDateOption(optionName: 'value' | 'min' | 'max'): Date | undefined | null | (Date | null)[] {
+    const { value } = this.option();
+    const isArray = optionName === 'value' && !this._isSingleMode();
+
+    if (!isArray) {
+      return this._convertToDate(this.option(optionName));
+    }
+
+    const valueArray = value ?? [];
+
+    return (valueArray as (DateLike | null)[]).map((item) => this._convertToDate(item));
```

Comment:
Use type guards instead; it could be implemented as follows:

  _isArrayValue(optionName: 'value' | 'min' | 'max', value: DateLike | DateLike[] | undefined): value is DateLike[] {
    return optionName === 'value' && !this._isSingleMode();
  }

  _getDateOption(optionName: 'value' | 'min' | 'max'): Date | undefined | null | (Date | null)[] {
    const { value } = this.option();

    if (!this._isArrayValue(optionName, value)) {
      const { [optionName]: optionValue } = this.option();

      return this._convertToDate(optionValue);
    }

    const valueArray = value ?? [];

    return valueArray.map((item) => this._convertToDate(item));
  }
  
  In this case typescript infers types correctly

Reply:
https://github.com/DevExpress/DevExtreme/commit/ec2f610d374bc1eb469a6ab73af78a85080ffa5d